pkg/inlinekeepalive/keepalive.go

```diff
@@ -0,0 +1,84 @@
+package inlinekeepalive
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/hashicorp/go-hclog"
+	"google.golang.org/protobuf/reflect/protoreflect"
+
+	pb "github.com/hashicorp/waypoint/pkg/server/gen"
+)
+
+const (
+	KeepaliveProtoSignature   = "inline_keepalive"
+	HeaderSendKeepalivesKey   = "wp-inline-keepalives"
+	HeaderSendKeepalivesValue = "true"
+	FeatureName               = "inline-keepalives"
+)
+
+// IsInlineKeepalive determines if a given proto message is
+// an inline keepalive.
+func IsInlineKeepalive(log hclog.Logger, m protoreflect.ProtoMessage) bool {
+	unknownFields := m.ProtoReflect().GetUnknown()
```

Comment:
Is there a reason we can't use the proto message directly here? If the keep alive proto message is always of a certain type we should be able to peek at it and determine if it's a keep alive.

```
keepAliveMsg, ok := m.(pb.InlineKeepalive)
if !ok { ... }
```

I think this would solve the field number reservation question I had earlier too.

Reply:
at that point, `m` is always the message type that the stream expects, like a `RunnerConfigResponse ` or an `EntrypointExecRequest`. The proto unmarshaller will have run by this point, and it will only ever return the expected type.

Reply:
I'm also confused about this - this logic seems like it's trying to handle the case where the server doesn't know about the message type, but surely it does?

Reply:
It is weird. I added a bunch of code comments, and a new github comment with a walkthrough video - if you still have questions, respond there!